docs/userguide/futures.rst

```diff
@@ -3,49 +3,80 @@
+
+Parsl supports concurrency and synchronization as follows. 
+Whenever a program calls a function that the programmer has defined to be a Parsl app (see :ref:`label-apps`),
+Parsl starts the function executing in a separate thread or process (on the same or another computer), returns a `future <https://en.wikipedia.org/wiki/Futures_and_promises>`_ in lieu of that function's result(s), 
+and continues immediately to the next statement in the Python program.
+Furthermore, when a computation attempts to access the value of a future (using mechanisms that we describe below), 
+Parsl blocks that computation until the future has a value. 
+
+More generaally, a future is an object that can be used track the status of an asynchronous task. 
+This object may, in the future, be interrogated to determine the task's status,
 results, exceptions, etc. A future is a proxy for a result that may not yet be available.
 
-Parsl provides two types of futures: AppFutures and DataFutures. While related, these two types of futures enable subtly different workflow patterns.
+Parsl provides two types of futures: AppFutures and DataFutures. While related, they enable subtly different workflow patterns.
 
 AppFutures
 ----------
 
-AppFutures are the basic building block upon which Parsl scripts are built. Every invocation of a Parsl app returns an AppFuture which may be used to manage execution and control the workflow.
+AppFutures are the basic building block upon which Parsl scripts are built. Every invocation of a Parsl app returns an AppFuture that may be used to monitor and manage the task's execution.
 AppFutures are inherited from Python's `concurrent library <https://docs.python.org/3/library/concurrent.futures.html>`_.
-AppFutures provide several key functionalities:
+They provide three key functionalities:
 
-1. An AppFuture provides a way to check the current status of an app.
+1. An AppFuture's ``result()`` function can be used to wait for an app to complete, and then access any result(s).
+This function is blocking: it returns only when the app completes or fails. 
+Thus the following code fragment implements essentially the behavior of the `expensive_function()` example above:
+the addition executes only when the two `sleep_double()` function calls complete.
 
    .. code-block:: python
 
-       @python_app
-       def double(x):
-             return x*2
+      @python_app
+      def sleep_double(x):
+           import time
+           time.sleep(2)   # Sleep for 2 seconds
+           return x*2
 
-       # doubled_x is an AppFuture
-       doubled_x = double(10)
+      # Start two concurrent sleep_double apps. doubled_x1 and doubled_x2 are AppFutures
+      doubled_x1 = sleep_double(10)
+      doubled_x2 = sleep_double(5)
 
-       # Check status of doubled_x, this will print True if the result is available, else false
-       print(doubled_x.done())
+      # The result() function will block until the two app calls have completed
```

Comment:
Until each of the corresponding app calls have completed? To avoid the user thinking result() is a global barrier.